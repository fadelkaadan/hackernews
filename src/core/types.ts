export interface IStory {
  by: string;
  descendants: number;
  id: number;
  kids: number[];
  score: number;
  text: string;
  time: number;
  title: string;
  type: string;
  url: string;
}

export interface IComment {
  by: string;
  id: number;
  kids: number[];
  parent: number;
  text: string;
  time: number;
  type: string;
}

export interface IUser {
  about: string;
  created: number;
  id: string;
  karma: number;
  submitted: number[];
}