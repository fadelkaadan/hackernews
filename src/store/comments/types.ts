export interface IComment {
  by: string;
  id: number;
  kids: number[];
  parent: number;
  text: string;
  time: number;
  type: string;
}

export interface CommentsState {
  list: IComment[];
  pending: Boolean;
  error: Object | null;
}

export const FETCH_COMMENTS_SUCCESS = "FETCH_COMMENTS_SUCCESS";
export const FETCH_COMMENTS_PENDING = "FETCH_COMMENTS_PENDING";
export const FETCH_COMMENTS_ERROR = "FETCH_COMMENTS_ERROR";

interface IFetchCommentsPendingAction {
  type: string;
}

interface IFetchCommentsSuccessAction {
  type: string;
  data: IComment[];
}

interface IFetchCommentsErrorAction {
  type: string;
  error: any;
}

export type CommentsActionTypes =
  | IFetchCommentsPendingAction
  | IFetchCommentsSuccessAction
  | IFetchCommentsErrorAction;
