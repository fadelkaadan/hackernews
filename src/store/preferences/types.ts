export enum ThemeEnum {
  DARK = "DARK",
  NIGHT = "NIGHT",
  LIGHT = "LIGHT",
}

export interface ThemeState {
  isPreferencesClicked: Boolean
  theme: ThemeEnum;
  isOneColumnLayout: Boolean;
}

export const TOGGLE_PREFERENCES = "TOGGLE_PREFERENCES";
export const SET_THEME = "SET_THEME";
export const TOGGLE_ONE_COLUMN_LAYOUT = "TOGGLE_ONE_COLUMN_LAYOUT";

interface togglePreferences {
  type: typeof TOGGLE_PREFERENCES;
}

interface setThemeAction {
  type: typeof SET_THEME;
  data: ThemeEnum;
}

interface toggleOneColumnLayoutAction {
  type: typeof TOGGLE_ONE_COLUMN_LAYOUT;
}

export type ThemeActionTypes =
  | setThemeAction
  | toggleOneColumnLayoutAction
  | togglePreferences;
