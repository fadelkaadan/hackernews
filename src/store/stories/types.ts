export interface IStory {
  by: string;
  descendants: number;
  id: number;
  kids: number[];
  score: number;
  text: string;
  time: number;
  title: string;
  type: string;
  url: string;
}

export interface StoriesState {
  list: IStory[] | undefined;
  pending: Boolean;
  error: Object | null;
}

export const FETCH_STORIES_SUCCESS = "FETCH_STORIES_SUCCESS";
export const FETCH_STORIES_PENDING = "FETCH_STORIES_PENDING";
export const FETCH_STORIES_ERROR = "FETCH_STORIES_ERROR";

interface fetchStoriesAction {
  type: string;
  data?: IStory[];
  error?: any;
}

export type StoriesActionTypes = fetchStoriesAction;
