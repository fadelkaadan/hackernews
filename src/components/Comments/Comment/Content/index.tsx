import React from "react";
import styled from "styled-components";
import { IComment } from "../../../../core/types";

interface ContentProps {
  data: IComment;
}

const Wrapper = styled.div`
  color: ${(props) => props.theme.main.text.primary};
  text-decoration: none;
  font-size: 1.3rem;
  width: 100%;
  margin: 0;
  line-height: 1.4;
  overflow-wrap: anywhere;

  & a {
    color: ${(props) => props.theme.main.text.primary};
  }

  & code {
    white-space: pre-line;
  }
`;

const Content = ({ data }: ContentProps) => {
  const stringToHTML = (str: string) => {
    var parser = new DOMParser();
    var doc = parser.parseFromString(str, "text/html");
    return doc.body.innerHTML;
  };

  return <Wrapper dangerouslySetInnerHTML={{ __html: stringToHTML(data.text) }} />;
};

export default Content;
