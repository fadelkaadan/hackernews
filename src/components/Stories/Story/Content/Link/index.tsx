import React, { useState } from "react";
import styled from "styled-components";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCopy } from "@fortawesome/free-regular-svg-icons";
import { faCheckCircle } from "@fortawesome/free-regular-svg-icons";
import copy from "copy-to-clipboard";

interface LinkProps {
  src: string;
}

const Wrapper = styled.div`
  cursor: pointer;
  margin-top: 0.6rem;
`;

const Button = styled.button`
  font-size: 1rem;
  padding: 0;
  cursor: pointer;
  background-color: transparent;
  border: none;
  color: ${(props) => props.theme.main.text.secondary};

  &:focus {
    outline: none;
    border: none;
  }

  &::after {
    content: "";
    display: block;
    width: 0%;
    padding-top: 0.3rem;
    border-bottom: 0.2rem solid #785bf7;
    transition: 0.2s;
  }

  &:hover::after {
    width: 100%;
  }
`;

const ButtonSuccess = styled(Button)`
  color: #83ad6c;
  &::after {
    border-bottom: 0.2rem solid #83ad6c;
  }

  &:hover::after {
    width: 0%;
  }
`;

const Link = ({ src }: LinkProps) => {
  const [icon, setIcon] = useState(faCopy);
  const [didCopy, setDidCopy] = useState(false);

  const getHostname = (url: string) => {
    const hostname = new URL(url).hostname.replace("www.", "");
    return hostname;
  };

  const addToClipboard = (text: string) => {
    copy(text);
    setIcon(faCheckCircle);
    setDidCopy(true);
    setTimeout(() => {
      setIcon(faCopy);
      setDidCopy(false);
    }, 1500);
  };

  const renderButton = () => {
    if (didCopy) {
      return (
        <ButtonSuccess onClick={() => addToClipboard(src)}>
          <FontAwesomeIcon icon={icon} /> Copied
        </ButtonSuccess>
      );
    }
    return (
      <Button onClick={() => addToClipboard(src)}>
        <FontAwesomeIcon icon={icon} /> {getHostname(src)}
      </Button>
    );
  };

  return <Wrapper>{renderButton()}</Wrapper>;
};

export default Link;
