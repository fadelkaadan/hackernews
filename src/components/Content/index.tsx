import React from "react";
import styled from "styled-components";
import TopInfo from "../TopInfo";
import Link from "../Link";
import Comments from "../Comments";

const Wrapper = styled.div`
  padding: 5px 15px 5px 15px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: flex-start;
`;


const Title = styled.a`
  color: ${(props) => props.theme.main.text.primary};
  text-decoration: none;

  &:visited {
    color: ${(props) => props.theme.main.text.disabled};
  }
`;

const StyledLink = styled.div`
  color: ${(props) => props.theme.main.text.primary};
`;

const Content = ({ story }: any) => {
  return (
    <Wrapper>
      <TopInfo time={story.time} score={story.score}/>
            <StyledLink>
              <Title href={story.url}>{story.title}</Title>
              <Link src={story.url} />
            </StyledLink>
            <Comments comments={story.kids} />
    </Wrapper>
  );
};

export default Content;
