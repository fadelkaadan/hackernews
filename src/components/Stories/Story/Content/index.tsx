import React from "react";
import styled from "styled-components";
import TopInfo from "../TopInfo";
import Link from "./Link";
import Comments from "../Comments";

interface IStory {
  by: string;
  descendants: number;
  id: number;
  kids: number[];
  score: number;
  text: string;
  time: number;
  title: string;
  type: string;
  url: string;
}

interface ContentProps {
  data: IStory;
}

const Wrapper = styled.div`
  padding: 5px 15px 5px 15px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: flex-start;
`;

const Title = styled.a`
  color: ${(props) => props.theme.main.text.primary};
  text-decoration: none;

  &:visited {
    color: ${(props) => props.theme.main.text.disabled};
  }
`;

const StyledLink = styled.div`
  color: ${(props) => props.theme.main.text.primary};
`;

const Content = ({ data }: ContentProps) => {
  return (
    <Wrapper>
      <TopInfo time={data.time} score={data.score} />
      <StyledLink>
        <Title href={data.url}>{data.title}</Title>
        <Link src={data.url} />
      </StyledLink>
      <Comments comments={data.kids} />
    </Wrapper>
  );
};

export default Content;
