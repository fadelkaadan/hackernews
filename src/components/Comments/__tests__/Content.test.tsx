import React from "react";
import { render, screen, RenderResult } from "../../../utilities/test-utils";
import Content from "../Content";
import { IComment } from "../../../core/types";

describe("Comment content", () => {
  const fakeContent = "Here is some content";
  const fakeContentHTML =
    "<p>Here <span>is</span> some content with <italic>html elements</italic></p>";

  const comment: IComment = {
    by: "fadel",
    id: 1,
    kids: [1, 2, 3],
    parent: 1,
    text: fakeContent,
    time: 1,
    type: "comment",
  };

  const commentHTML: IComment = {
    by: "fadel",
    id: 1,
    kids: [1, 2, 3],
    parent: 1,
    text: fakeContentHTML,
    time: 1,
    type: "comment",
  };

  it("Should match the snapshot", () => {
    const { container }: RenderResult = render(<Content data={comment} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should render normal text", () => {
    render(<Content data={comment} />);
    const content: HTMLElement = screen.getByText(fakeContent);
    expect(content.textContent).toEqual(fakeContent);
  });

  it("Should render html elements", () => {
    const { container }: RenderResult = render(<Content data={commentHTML} />);
    const span = container.querySelector("span");
    expect(span).toBeDefined();
  });
});
