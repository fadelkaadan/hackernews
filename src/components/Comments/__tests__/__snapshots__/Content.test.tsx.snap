// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment content Should match the snapshot 1`] = `
<div
  class="sc-bdnylx iPnIIX"
>
  Here is some content
</div>
`;
